### PR TITLE
metadata-store: modernisation

### PR DESCRIPTION
**Pull request by liam-fitzgerald**
- Switches metadata-store to resource identifiers
- The core `associations` datastructure is now a `(map md-resource [group=resource =metadata])`
- The indices are cached, a la hark, so as to simplify OTA migrations in future
- Deprecates metadata-hook, switches to the push/pull libraries
- Only admins and moderators can add channels now

#### TODO:

- [x] OTA logic
- [x] Join Group preview
- [x] Think about import of metadata-hook
- [x] Allow host to configure channels in preview

### CODE REVIEW

File: pkg/interface/src/views/apps/publish/components/MetadataForm.tsx [hunk 1]
```diff
@@ -64,11 +64,11 @@ export function MetadataForm(props: MetadataFormProps) {
...
       await api.metadata.metadataAdd(
         "publish",
-        props.association["app-path"],
-        props.association["group-path"],
+        props.association.resource,
+        props.association.group,
         name,
         description,
-        props.association.metadata["date-created"],
+        props.association.metadata["date-created"],,
```
Comment (tacryt-socryp) [inline pkg/interface/src/views/apps/publish/components/MetadataForm.tsx:71]: Does this compile?

File: pkg/arvo/app/metadata-store.hoon [hunk 2]
```diff
@@ -44,11 +44,35 @@
...
+::
++$  cached-indices
+  $:  group-indices=(jug resource md-resource)
+      app-indices=(jug app-name [group=resource =resource])
+      resource-indices=(map md-resource resource)
+  ==
+::
++$  base-state-2
+  $:  =associations
+      ~
```
Comment (tacryt-socryp) [inline pkg/arvo/app/metadata-store.hoon:75]: Why a sig?

File: pkg/interface/src/views/apps/publish/components/MetadataForm.tsx (hunk 1 above)
Comment (liam-fitzgerald) [reply pkg/interface/src/views/apps/publish/components/MetadataForm.tsx:71]: Hahah, just realised this component is no longer used. Removing.

File: pkg/arvo/app/metadata-store.hoon (hunk 2 above)
Comment (liam-fitzgerald) [reply pkg/arvo/app/metadata-store.hoon:75]: idk, something makes me mildly uncomfortable about a base-state that's just an alias for a type. Happy to change if you feel strongly about this

**Comment by liam-fitzgerald** [comment]
Added a JoinGroup preview. Still waiting on designs for allowing hosts to configure the channels that show in the preview. Sorting by popularity is a non-starter because it would require the host to be subscribed to every graph, would probably bring the larger groups to their knees. 
Capture d’écran 2021-01-15 à 3 46 31 PM

**Comment by liam-fitzgerald** [comment]
Also requesting clarification on the behaviour when the group host is down. We should be able to force a join even if the host is down, but we won't be able to get a nice preview for it.

**Comment by matildepark** [comment]
How would we force a join? We should just say it’s unavailable to join if we can’t reach it no?

**Comment by liam-fitzgerald** [comment]
just send the join request and then it'll get there eventually. but realistically group owners should be always on, so it's not really that much of an issue

**Comment by liam-fitzgerald** [comment]
@belisarius222 with the pass-through watch pattern in the metadata-hooks, how should offline ships be handled? Would a simple timeout suffice? e.g. if we don't hear back in some period then inform landscape and hang up the subscription?

**Comment by liam-fitzgerald** [comment]
Capture d’écran 2021-01-18 à 12 03 16 PM

manual exposure of preview channels is done, along with channel removal from the group
Capture d’écran 2021-01-18 à 12 14 17 PM

This is ready for the big leagues

**Comment by jfranklin9000** [comment]
The 'g' and 'p' are clipped in "Remove from group".

**Comment by urcades** [comment]
Checking this out now — will begin to append any design feedback in the form of screenshots/commentary here

**Comment by urcades** [comment]
receiving a weird stack trace the moment I log into the ship running this PR's code:


```
TypeError: Cannot read property 'substr' of undefined
    at http://0.0.0.0:9000/~landscape/js/bundle/index.86c6e416c338a305e1e9.js:201376:37
    at Array.map (<anonymous>)
    at http://0.0.0.0:9000/~landscape/js/bundle/index.86c6e416c338a305e1e9.js:201356:34
    at Array.map (<anonymous>)
    at index (http://0.0.0.0:9000/~landscape/js/bundle/index.86c6e416c338a305e1e9.js:201354:6)
    at Omnibox.componentDidUpdate (http://0.0.0.0:9000/~landscape/js/bundle/index.86c6e416c338a305e1e9.js:223749:93)
    at Omnibox.componentDidUpdate (http://0.0.0.0:9000/~landscape/js/bundle/index.86c6e416c338a305e1e9.js:161165:123)
    at commitLifeCycles (http://0.0.0.0:9000/~landscape/js/bundle/index.86c6e416c338a305e1e9.js:152763:22)
    at commitLayoutEffects (http://0.0.0.0:9000/~landscape/js/bundle/index.86c6e416c338a305e1e9.js:155731:7)
    at HTMLUnknownElement.callCallback (http://0.0.0.0:9000/~landscape/js/bundle/index.86c6e416c338a305e1e9.js:133116:14)
    ```

**Comment by urcades** [comment]
I'm unable to proceed past this error – it's blocking my ability to visit notifications to check out a group invite, or open up leap for that matter!

**Comment by matildepark** [comment]
You are running this against a fake zod right? This is not compatible with livenet at all.

**Comment by urcades** [comment]
@matildepark Ha, shoot – didn't know that was a limitation, ty!

**Comment by matildepark** [comment]
@urcades just booting a fakezod, eg. `./urbit-v1.0-x86_64-darwin/urbit -F zod -B "bin/solid.pill" -A "pkg/arvo"` (if you untarred the install binary in your urbit repo folder as I did), will produce the metadata modernisation work on `localhost:80`.

@liam-fitzgerald Two issues I found while testing this for Ed was that creating a channel doesn't auto-redirect to the channel anymore and metadata-push-hook et al aren't auto-started in drum.

**Comment by urcades** [comment]
@matildepark  + @liam-fitzgerald  happy news! I finally got this working, with two ships able to send one another group invites.

I see the addition of "Channels" to the group's settings, which is a welcome change.

How do I trigger the display of the "Join Group" modal? When I tried to click through a group link, or accept the invite via Notifications, I was simply taken to the group's index/root directly.

**Comment by urcades** [comment]
Besides not understanding how to trigger the join modal, the elements of metadata that are surfaced, are of themselves, looking good.

Some notes on later implementation/design, as we plug this into some of the forthcoming Settings reworks:

Screen Shot 2021-01-21 at 5 02 28 PM

We should probably specify further in the dropdown copy what exactly is being pinned to what — the copy in the dropdown states "Unpin from Join Card" — perhaps we'll develop a more elegant descriptor for the "join group interstitial",  but this is fine for now.

Note some of the existing actions I've added to this dropdown are reminiscent of the actions we afford from the channel itself — that's because this is the case. This will be more crucial, again, as "channel management" is funneled into its own interface.

Apart from these small changes, given I can't actually view the join group card on my local ships, the above screenshot looks pretty representative of our current designs.

**Comment by liam-fitzgerald** [comment]
ready for review @tacryt-socryp

**Comment by liam-fitzgerald** [comment]
ready for review @tacryt-socryp

File: pkg/arvo/app/metadata-push-hook.hoon [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  ::
+  ++  channels
+    %-  ~(gas by *associations)
+    %+  skim  ~(tap by (app-metadata-for-group:met rid %graph))
+    |=([=md-resource group=resource =metadata] preview.metadata)
+  ::   
+  ++  channel-count
+    ~(wyt by (app-metadata-for-group:met rid %graph))
+  --
+
```
Comment (tacryt-socryp) [inline pkg/arvo/app/metadata-push-hook.hoon:68]: blank line

File: pkg/arvo/sur/metadata-store.hoon [hunk 4]
```diff
@@ -1,28 +1,61 @@
...
       =color
       date-created=time
       creator=ship
       module=term
+      picture=url
+      preview=?
+      vip=vip-metadata
   ==
 ::
 +$  metadata-action
```
Comment (tacryt-socryp) [inline pkg/arvo/sur/metadata-store.hoon:45]: Name should be changed (modernized) to either `action` or `update`.

File: pkg/arvo/sur/metadata-store.hoon [hunk 5]
```diff
@@ -1,28 +1,61 @@
...
+      [%remove group=resource resource=md-resource]
+      [%initial-group group=resource =associations]
   ==
 ::
++$  metadata-hook-update
+   $%  [%req-preview group=resource]
+       [%preview group-preview]
+   ==
+::
 +$  metadata-update
```
Comment (tacryt-socryp) [inline pkg/arvo/sur/metadata-store.hoon:56]: Name should change to `update`

File: pkg/arvo/sur/metadata-store.hoon [hunk 6]
```diff
@@ -1,28 +1,61 @@
...
 ::
 +$  metadata-action
-  $%  [%add =group-path resource=md-resource =metadata]
-      [%remove =group-path resource=md-resource]
+  $%  [%add group=resource resource=md-resource =metadata]
+      [%remove group=resource resource=md-resource]
+      [%initial-group group=resource =associations]
   ==
 ::
++$  metadata-hook-update
```
Comment (tacryt-socryp) [inline pkg/arvo/sur/metadata-store.hoon:51]: Name should change to `hook-update`

File: pkg/arvo/lib/metadata-json.hoon [hunk 7]
```diff
@@ -1,24 +1,26 @@
 /-  *metadata-store
+/+  resource
+^?
```
Comment (tacryt-socryp) [inline pkg/arvo/lib/metadata-json.hoon:3]: File name should be changed from `metadata-json` to `metadata-store` to match the other agent libraries. We should also update this file while we're at it to follow the `enjs` and `dejs` pattern seen elsewhere.

File: pkg/arvo/app/metadata-store.hoon (hunk 2 above)
Comment (tacryt-socryp) [reply pkg/arvo/app/metadata-store.hoon:75]: Nah, doesn't matter

File: pkg/arvo/app/metadata-store.hoon [hunk 8]
```diff
@@ -192,293 +207,227 @@
...
-        app-indices
-      (rebuild-app-indices associations)
+         resource-indices
+       (rebuild-resource-indices associations.old)
       ::
-        group-indices
-      (rebuild-group-indices associations)
+         group-indices
+      (rebuild-group-indices associations.old)
+      ::
```
Comment (tacryt-socryp) [inline pkg/arvo/app/metadata-store.hoon:222]: dedent the comments to be in line with the `%_`

**Review by tacryt-socryp — APPROVED** [review]
The logic of this PR is great. Very clear and cogent. A few style nitpicks and a few parts of the libs that were not fully modernized. LGTM, can merge once the remaining naming issues are made consistent with the new style.

**Pull request merged**